aeneas/src/ir/PackingSolver.v3

```diff
@@ -0,0 +1,332 @@
+			for (i < elements.length) elements[i] = true;
+			result = canDistinguish(state, elements);
+		}
+		return result;
+	}
+
+	// Backtracking solver
+	def solve(idx: int, state: Array<PackingState>) -> HashMap<FieldLoc, Interval> {
+		// for (i < state.length) Terminal.put3("%d %d: %q\n", idx, i, state[i].render);
+		if (idx == fieldOrder.length) return if(checkDistinguishable(state), assignments, null);
+
+		var fieldLoc = fieldOrder[idx];
+		var caseIdx = fieldLoc.caseIdx, fieldIdx = fieldLoc.fieldIdx;
+		var field = cases[caseIdx][fieldIdx];
+		var possibleIntervals = state[caseIdx].getIntervalsForSize(PackingField.Nonref.!(field).size);
+		if (possibleIntervals.length == 0) return null;
+
+		for (interval in possibleIntervals) {
+			state[caseIdx].assignInterval(interval);
+			assignments[fieldLoc] = interval;
+			var result = solve(idx + 1, state);
+			if (result != null) return result;
+			state[caseIdx].unassignInterval(interval);
+		}
+		return null;
+	}
+
+	// Find a packing for a single scalar
+	def solveScalar(cases: Array<Array<PackingField>>, preassigned: Array<(FieldLoc, Interval)>) -> HashMap<FieldLoc, Interval> {
+		var isRefScalar = false;
+		for (i < cases.length) {
+			var caseHasRef = false;
+			for (field in cases[i]) {
+				match (field) {
+					Ref => {
+						if (caseHasRef) return null; // cannot pack more than one ref into a single scalar per case
+						isRefScalar = caseHasRef = true;
+					}
+					_ => ;
+				}
+			}
+		}
+
+		var initialState = Array<PackingState>.new(cases.length);
+		if (isRefScalar) {
+			for (i < cases.length) {
+				var c = cases[i];
+				var casePacking: PackingState;
+
+				var containsRef = false;
+				for (j < cases[i].length) {
+					var f = cases[i][j], fl = FieldLoc(i, j);
+					match (f) {
+						Ref => {
+							assignments[fl] = gcPatterns.refInterval;
+							casePacking = gcPatterns.scanned.copy();
+							casePacking.assignInterval(gcPatterns.refInterval);
+							containsRef = true;
+						}
+						_ => ;
+					}
+				}
+
+				if (!containsRef) casePacking = gcPatterns.ignored.copy();
+				initialState[i] = casePacking;
+			}
+		} else {
+			// can use all bits
+			var basePacking = PackingState.new(Array<Bit>.new(size));
+			for (i < cases.length) initialState[i] = basePacking.copy();
+		}
+
+		for (assignment in preassigned) {
+			var fl = assignment.0, interval = assignment.1;
+			initialState[fl.caseIdx].assignInterval(interval);
+			assignments[fl] = interval;
+		}
+
+		var fieldsToSolve = Vector<FieldLoc>.new();
+		for (i < cases.length) {
+			for (j < cases[i].length) {
+				var fl = FieldLoc(i, j);
+				if (!assignments.has(fl)) fieldsToSolve.put(fl);
+			}
+		}
+		fieldOrder = fieldsToSolve.extract();
+		this.cases = cases;
+		
+		var solution = solve(0, initialState);
+		if (solution == null) return null;
+
+		return assignments;
+	}
+}
+
+type GcPatterns(scanned: PackingState, refInterval: Interval, ignored: PackingState);
```

Comment:
What about the terms `ptrref` and `nonptrref`? Both are reference words (i.e. marked as a reference in stackmaps and object maps), but `ptrref`s contain addresses that are followed by the GC.